src/common/atomic_queue.cpp

```diff
@@ -0,0 +1,397 @@
+        message << "Atomic Queue "<< qId << " Full";
+        pthread_mutex_unlock(&pushQmutex[qId]);
+        throw Memory_Service_Exception(ATOMIC_QUEUE_FULL,message.str().c_str()); //array full
+    }
+    // Incramenet the rear and size
+    lcqData->rear = (lcqData->rear + 1) % lcqData->capacity;
+    lcqData->size.fetch_add(1,std::memory_order_relaxed);
+    openfam_persist(pointerQ,sizeof(qData));
+    if (lcqData->size.load(std::memory_order_relaxed) == 1)
+        pthread_cond_signal(&empty[qId]); // Signal to processing thread that queue is not empty anymore
+    pthread_mutex_unlock(&pushQmutex[qId]);
+    return 0;
+}
+/*Pop element from the queue using front pointer
+* @param item - Fam_Global_Descriptor
+* @return - {success(0), failure(1),ATOMIC_QUEUE_EMPTY
+*/
+int atomicQueue::read(Fam_Global_Descriptor *item) {
+    struct qData *lcqData;
+    lcqData = (struct qData *)pointerQ;
+    if ( lcqData->size.load(std::memory_order_relaxed) == 0 )
+        return ATOMIC_QUEUE_EMPTY; //array empty
+    item->regionId = ATOMIC_REGION_ID;
+    // front should hold the offset of the data to be read
+    item->offset = *((uint64_t *)pointerA+lcqData->front);
+    return 0;
+}
+/* Remove the element from queue which was poped
+* @param item - Fam_Global_Descriptor
+* @return - {success(0), failure(1)}
+*/
+
+int atomicQueue::pop(Fam_Global_Descriptor *item) {
+    void * localPointerM;
+    struct qData *lcqData;
+    lcqData = (struct qData *)pointerQ;
+    // Adjust the front and size
+    lcqData->front = (lcqData->front + 1) % lcqData->capacity;
+    lcqData->size.fetch_add(-1, std::memory_order_relaxed);
+    openfam_persist(pointerQ,sizeof(qData));
+    localPointerM = allocator->get_local_pointer(item->regionId,item->offset);
+    //clear the content of the message
+    memset(localPointerM,0,sizeof(atomicMsg));
+    openfam_persist(localPointerM,sizeof(atomicMsg));
+    return 0;
+}
+/*Check whether the queue is empty
+* return true(1), false(0)
+*/
+bool atomicQueue::isQempty() {
+    struct qData *lcqData;
+    lcqData = (struct qData *)pointerQ;
+    return ((lcqData->size == 0) ? true : false);
+}
+/*Recover the incomplete transactions during startup
+* @param qId - queue number
+* @param - Memserver_Allocator
+*/
+int recover_queue(uint32_t qId, Memserver_Allocator * allocator) {
+    Fam_Global_Descriptor item;
+    void *localPointer, *localPointerB, *localPointerD;
+    atomicMsg *msgPointer;
+    while (!atomicQ[qId].isQempty()) {
+        atomicQ[qId].read(&item);
+        localPointer = allocator->get_local_pointer(item.regionId,item.offset);
+        msgPointer = (atomicMsg *)localPointer;
+        if (msgPointer->flag == ATOMIC_READ) {
+        // Atomic read; just remove the element from the queue; don't perform any action
+            atomicQ[qId].pop(&item);
+            continue;
+        }
+        if ((msgPointer->flag & ATOMIC_WRITE) && (msgPointer->flag & ATOMIC_WRITE_IN_PROGRESS)) {
```

Comment:
Possible rearrangement of code here ? Only for ATOMIC_WRITE_IN_PROGRESS, we need some processing in recover_queue. Every other condition we just need to pop the item. May be you can just remove "(msgPointer->flag == ATOMIC_READ) " ??